0008-security-recommendations/README.md

```diff
@@ -0,0 +1,98 @@
+
+The general system consists of a server, clients, and end devices. End devices are devices that produce data, and the client only serves as a proxy for them. End devices are represented as endpoints within 1/KP.
+
+So the connection scheme looks like this:
+```
+server -- clients -- end devices
+```
+
+All data flows between the server and end devices via clients.
+
+As you see, there are two links involved: server-client and client-device. These links are purely logical. (e.g., there is no physical link if the client acts as an end device and produces data on its own.)
+
+## Goals
+The goal of this document is to ensure _confidentiality_, _integrity_, and _authenticity_ of data flowing between the server and end devices.
+
+_Confidentiality_ "is the property, that information is not made available or disclosed to unauthorized individuals, entities, or processes" (Excerpt ISO27000).
+
+_Data integrity_ means maintaining and assuring the accuracy and completeness of data over its entire life-cycle. This means that data cannot be modified in an unauthorized or undetected manner.
+
+_Authenticity_ is the property that ensures that the identity of a subject or resource is the identity claimed.
+
+_Note: in general, you want to authenticate both servers, clients, and devices._
+
+The whole system is secure when all three properties are ensured for both links.
+
+Note also that not all properties might be required for the system at the same time, but authenticity usually requires data integrity, and often relies on confidentiality to be effective.
+
+## Possible solutions
+This section describes possible solutions for ensuring necessary properties for both links.
+
+Note that the easiest way to ensure all properties at the same time is making the link purely virtual. That could be achieved by placing both ends of a link on the same software or hardware system.
+
+Examples are running both client and end devices on the same hardware machine, or running both client and server on the same host and communicate over loop interface. (While that is not extremely practical in IoT solution it well may happen during development.)
+
+### Server-Client
+Confidentiality:
+
+- channel encryption (provided by TLS)
+- MQTT payload encryption to server/client key
+
+Integrity:
+
+- Message Authentication Code (MAC). e.g., HMAC (provided by TLS)
+- payload signatures with server/client key
+
+Authenticity:
+
+- certificate authentication (provided by TLS)
+- MQTT username/password (this does not authenticate the server)
+
+When using a bi-directional certificate authentication, it is recommended to issue separate certificates for each client instance, to limit the effect of private key stealing. Unfortunately, that is not always possible (e.g., mass production device); in that case, it is recommended to pass some unique client identifier using MQTT username/password (and using whitelists).
```

Comment:
Well, even if one MQTT login/pass gets compromised, that does not affect other clients. A single compromised cert applied to all clients ruins security for all of them.